pkg/sql/plan/build_dml_util.go

```diff
@@ -2174,6 +2234,192 @@ func appendPreInsertNode(builder *QueryBuilder, bindCtx *BindContext,
 	return lastNodeId
 }
 
+func appendPreInsertSkVectorPlan(
+	builder *QueryBuilder,
+	bindCtx *BindContext,
+	tableDef *TableDef,
+	lastNodeId int32,
+	multiTableIndex *MultiTableIndex,
+	isUpdate bool,
+	idxRefs []*ObjectRef,
+	indexTableDefs []*TableDef) (int32, error) {
+
+	//1. get vec & pk column
+	colsMap := make(map[string]int)
+	colTypes := make([]*Type, len(tableDef.Cols))
+	for i, col := range tableDef.Cols {
+		colsMap[col.Name] = i
+		colTypes[i] = tableDef.Cols[i].Typ
+	}
+
+	var posOriginPk, posOriginVecColumn int
+	var typeOriginPk, typeOriginVecColumn *Type
+	varcharType := types.T_varchar.ToType()
+	bigIntType := types.T_int64.ToType()
+	{
+		for _, part := range multiTableIndex.IndexDefs[catalog.SystemSI_IVFFLAT_TblType_Entries].Parts {
+			if i, ok := colsMap[part]; ok {
+				posOriginVecColumn = i
+				typeOriginVecColumn = tableDef.Cols[i].Typ
+				break
+			}
+		}
+
+		posOriginPk, typeOriginPk = getPkPos(tableDef, false)
+	}
+	fmt.Println(posOriginVecColumn)
+	fmt.Println(typeOriginVecColumn)
+	fmt.Println(posOriginPk)
+	fmt.Println(typeOriginPk)
+	fmt.Println(bigIntType)
+
+	var leftChildId = lastNodeId
+	var rightChildId int32
+	{
+		scanNodeProject := make([]*Expr, len(indexTableDefs[1].Cols))
+		for colIdx, column := range indexTableDefs[1].Cols {
+			scanNodeProject[colIdx] = &plan.Expr{
+				Typ: column.Typ,
+				Expr: &plan.Expr_Col{
+					Col: &plan.ColRef{
+						ColPos: int32(colIdx),
+						Name:   column.Name,
+					},
+				},
+			}
+		}
+		rightChildId = builder.appendNode(&Node{
+			NodeType:    plan.Node_TABLE_SCAN,
+			Stats:       &plan.Stats{},
+			ObjRef:      idxRefs[1],
+			TableDef:    indexTableDefs[1],
+			ProjectList: scanNodeProject,
+		}, bindCtx)
+	}
+
+	var joinID int32
+	{
+
+		centroidsCentroid := &plan.Expr{
+			Typ: typeOriginVecColumn,
+			Expr: &plan.Expr_Col{
+				Col: &plan.ColRef{
+					RelPos: 1,
+					ColPos: 2,
+					Name:   catalog.SystemSI_IVFFLAT_TblCol_Centroids_centroid,
+				},
+			},
+		}
+		inputEmbedding := &plan.Expr{
+			Typ: typeOriginVecColumn,
+			Expr: &plan.Expr_Col{
+				Col: &plan.ColRef{
+					RelPos: 0,
+					ColPos: int32(posOriginVecColumn),
+					Name:   tableDef.Cols[posOriginVecColumn].Name,
```

Comment:
Getting stuck here!
ERROR 20101 (HY000): internal error: decode slice that is not a multiple of element size

How to use RelPos for `Order By l2_distance(col1, col2)` JOIN?